### ISSUE TITLE
tight_layout relies on get_renderer, which is not available on all backends

### ISSUE THREAD

**Issue by mdehoon** [matplotlib/matplotlib#1852]
With the MacOSX backend and possibly other backends, tight_layout falls back to the Agg backend. For example, demo_tight_layout.py does the following:

$ python -i examples/pylab_examples/demo_tight_layout.py
/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/matplotlib-1.3.x-py2.7-macosx-10.6-intel.egg/matplotlib/tight_layout.py:225: UserWarning: tight_layout : falling back to Agg renderer
  warnings.warn("tight_layout : falling back to Agg renderer")

This is the relevant code in tight_layout.py:

```
def get_renderer(fig):
    if fig._cachedRenderer:
        renderer = fig._cachedRenderer
    else:
        canvas = fig.canvas  
        if canvas and hasattr(canvas, "get_renderer"):
            renderer = canvas.get_renderer()
        else:
            # not sure if this can happen
            warnings.warn("tight_layout : falling back to Agg renderer")
            from matplotlib.backends.backend_agg import FigureCanvasAgg
            canvas = FigureCanvasAgg(fig)
            renderer = canvas.get_renderer()
    return renderer
```

Neither backend_template.py nor backend_bases.py implement a get_renderer method, nor do the MacOSX and gtkcairo backend (and possibly other backends). 

Even if I were to add a get_renderer method to the canvas in the MacOSX backend, this still fails. The problem is that this code tries to grab a renderer and use it outside of the event loop (i.e., outside of the figure.draw method). Outside of the event loop, the graphics context is not defined, and the code fails as follows:

$ python -i examples/pylab_examples/demo_tight_layout.py
Traceback (most recent call last):
File "examples/pylab_examples/demo_tight_layout.py", line 15, in
plt.tight_layout()
...
RuntimeError: CGContextRef is NULL

I would suggest to restructure the code in tight_layout.py to avoid using the renderer directly.

**Comment by aflansburg** [comment]
@c-benko Error is present on my Mac OSX 10.10.1 as well. Python 2.7.9, matplotlib 1.4.0 backend "MacOSX"

```python
/Users/myName/anaconda/lib/python2.7/site-packages/matplotlib/tight_layout.py:225: UserWarning: tight_layout : falling back to Agg renderer
  warnings.warn("tight_layout : falling back to Agg renderer")
```
Using
```python
from matplotlib import pyplot as plt
```
and calling
```python
plt.tight_layout()
plt.show
```

**Comment by diego898** [comment]
I am seeing this warning while running anything using pycharm and anaconda with python 3.5.2 and matplotlib 1.5.3.

I dont see this if I run through the jupyter notebook.

**Comment by janssen** [comment]
I'm seeing this with 2.0.0rc1 and the Kivy backend.

**Comment by bayesfactor** [comment]
I'm having this issue with SVG backend on Windows + python 3.5.5 + matplotlib 2.1.1

**Comment by smithsp** [comment]
Bump for PDF backend.

**Comment by anntzer** [comment]
Closed by #15221, I think.
